ignite/trainer.py

```diff
@@ -0,0 +1,151 @@
+            if training_step_result is not None:
+                self.training_history.append(training_step_result)
+
+            self.current_iteration += 1
+
+            self._fire_event(TrainingEvents.TRAINING_ITERATION_COMPLETED)
+            if self.should_terminate:
+                return
+
+        time_taken = time.time() - start_time
+        hours, mins, secs = _to_hours_mins_secs(time_taken)
+        self._logger.info("Epoch[%s] Complete. Time taken: %02d:%02d:%02d", self.current_epoch, hours,
+                          mins, secs)
+
+    def run(self, training_data, max_epochs=1):
+        """
+        Train the model, evaluate the validation set and update best parameters if the validation loss
+        improves.
+        In the event that the validation set is not run (or doesn't exist), the training loss is used
+        to update the best parameters.
+
+        Parameters
+        ----------
+        training_data : Iterable
+            Collection of training batches allowing repeated iteration (e.g., list or DataLoader)
+        max_epochs: int, optional
+            max epochs to train for [default=1]
+
+        Returns
+        -------
+        None
+        """
+
+        try:
+            self._logger.info("Training starting with max_epochs={}".format(max_epochs))
+
+            self.max_epochs = max_epochs
+
+            start_time = time.time()
+
+            self._fire_event(TrainingEvents.TRAINING_STARTED)
+            while self.current_epoch < max_epochs and not self.should_terminate:
+                self._fire_event(TrainingEvents.TRAINING_EPOCH_STARTED)
+                self._train_one_epoch(training_data)
+                if self.should_terminate:
+                    break
+                self._fire_event(TrainingEvents.TRAINING_EPOCH_COMPLETED)
+                self.current_epoch += 1
+
+            self._fire_event(TrainingEvents.TRAINING_COMPLETED)
+            time_taken = time.time() - start_time
+            mins, secs = divmod(time_taken, 60)
+            hours, mins = divmod(mins, 60)
+            self._logger.info("Training complete. Time taken %02d:%02d:%02d" % (hours, mins, secs))
+        except BaseException as e:
+            self._logger.error("Training is terminating due to exception: %s", str(e))
+            self._fire_event(TrainingEvents.EXCEPTION_RAISED)
+            raise e
+
+
+def create_supervised(model, optimizer, loss_fn, cuda=False):
+    """
+    Factory function for creating a trainer for supervised models
+
+    Args:
+        model (torch.nn.Module): the model to train
+        optimizer (torch.optim.Optimizer): the optimizer to use
+        loss_fn (torch.nn loss function): the loss function to use
+        cuda (bool, optional): whether or not to transfer batch to GPU (default: False)
+
+    Returns:
+        Trainer: a trainer instance with supervised update function
+    """
+    def _prepare_batch(batch, volatile=False):
```

Comment:
there is no way to access `volatile` parameter, i believe it should be exposed to the outer function.

Reply:
We usually can't train models with `volatile=True`, so I think it's ok not to expose it here.
BTW, with latest pytorch `volatile` was replaced by `with torch.no_grad()` context manager.

Reply:
Then we can remove the parameter just as well, right?

Reply:
I think we should just remove the param. This is a hangover from the refactor (copy-pasting)

Reply:
removed the param